OHOS: Use message-queue syscalls without triggering unsupported host syscalls.
--- a/linux-user/syscall.c
+++ b/linux-user/syscall.c
@@ -1302,7 +1302,18 @@
 
 #if defined(TARGET_NR_mq_open) && defined(__NR_mq_open)
 #include <mqueue.h>
+#ifdef __OHOS__
+#include <sys/utsname.h>
 
+static bool host_mq_supported(void)
+{
+    struct utsname host;
+
+    /* The HarmonyOS PC kills message-queue syscalls with SIGSYS. */
+    return uname(&host) == 0 && strcmp(host.sysname, "HarmonyOS") != 0;
+}
+#endif
+
 static inline abi_long copy_from_user_mq_attr(struct mq_attr *attr,
                                               abi_ulong target_mq_attr_addr)
 {
@@ -13667,6 +13678,11 @@
 
 #if defined(TARGET_NR_mq_open) && defined(__NR_mq_open)
     case TARGET_NR_mq_open:
+#ifdef __OHOS__
+        if (!host_mq_supported()) {
+            return -TARGET_ENOSYS;
+        }
+#endif
         {
             struct mq_attr posix_mq_attr;
             struct mq_attr *pposix_mq_attr;
@@ -13684,7 +13700,12 @@
             if (!p) {
                 return -TARGET_EFAULT;
             }
+#ifdef __OHOS__
+            ret = get_errno(syscall(__NR_mq_open, (char *)p + (*(char *)p == '/'), host_flags,
+                                    arg3, pposix_mq_attr));
+#else
             ret = get_errno(mq_open(p, host_flags, arg3, pposix_mq_attr));
+#endif
             unlock_user (p, arg1, 0);
         }
         return ret;
@@ -13691,5 +13712,10 @@
 
     case TARGET_NR_mq_unlink:
+#ifdef __OHOS__
+        if (!host_mq_supported()) {
+            return -TARGET_ENOSYS;
+        }
+#endif
         p = lock_user_string(arg1 - 1);
         if (!p) {
             return -TARGET_EFAULT;
@@ -13696,5 +13722,9 @@
         }
+#ifdef __OHOS__
+        ret = get_errno(syscall(__NR_mq_unlink, (char *)p + (*(char *)p == '/')));
+#else
         ret = get_errno(mq_unlink(p));
+#endif
         unlock_user (p, arg1, 0);
         return ret;
 
@@ -13701,5 +13731,10 @@
 #ifdef TARGET_NR_mq_timedsend
     case TARGET_NR_mq_timedsend:
+#ifdef __OHOS__
+        if (!host_mq_supported()) {
+            return -TARGET_ENOSYS;
+        }
+#endif
         {
             struct timespec ts;
 
@@ -13721,6 +13756,11 @@
 #endif
 #ifdef TARGET_NR_mq_timedsend_time64
     case TARGET_NR_mq_timedsend_time64:
+#ifdef __OHOS__
+        if (!host_mq_supported()) {
+            return -TARGET_ENOSYS;
+        }
+#endif
         {
             struct timespec ts;
 
@@ -13743,6 +13783,11 @@
 
 #ifdef TARGET_NR_mq_timedreceive
     case TARGET_NR_mq_timedreceive:
+#ifdef __OHOS__
+        if (!host_mq_supported()) {
+            return -TARGET_ENOSYS;
+        }
+#endif
         {
             struct timespec ts;
             unsigned int prio;
@@ -13769,6 +13814,11 @@
 #endif
 #ifdef TARGET_NR_mq_timedreceive_time64
     case TARGET_NR_mq_timedreceive_time64:
+#ifdef __OHOS__
+        if (!host_mq_supported()) {
+            return -TARGET_ENOSYS;
+        }
+#endif
         {
             struct timespec ts;
             unsigned int prio;
@@ -13800,6 +13850,11 @@
 /*         break; */
 
     case TARGET_NR_mq_getsetattr:
+#ifdef __OHOS__
+        if (!host_mq_supported()) {
+            return -TARGET_ENOSYS;
+        }
+#endif
         {
             struct mq_attr posix_mq_attr_in, posix_mq_attr_out;
             ret = 0;
@@ -13806,9 +13861,19 @@
             if (arg2 != 0) {
                 copy_from_user_mq_attr(&posix_mq_attr_in, arg2);
+#ifdef __OHOS__
+                ret = get_errno(syscall(__NR_mq_getsetattr, arg1,
+                                        &posix_mq_attr_in, &posix_mq_attr_out));
+#else
                 ret = get_errno(mq_setattr(arg1, &posix_mq_attr_in,
                                            &posix_mq_attr_out));
+#endif
             } else if (arg3 != 0) {
+#ifdef __OHOS__
+                ret = get_errno(syscall(__NR_mq_getsetattr, arg1, NULL,
+                                        &posix_mq_attr_out));
+#else
                 ret = get_errno(mq_getattr(arg1, &posix_mq_attr_out));
+#endif
             }
             if (ret == 0 && arg3 != 0) {
                 copy_to_user_mq_attr(arg3, &posix_mq_attr_out);
